presto-orc/src/main/java/com/facebook/presto/orc/reader/ApacheHiveTimestampDecoder.java

```diff
@@ -36,8 +38,32 @@ public static long decodeTimestamp(long seconds, long serializedNanos, DecodeTim
         if (value < 0 && nanos != 0) {
             value -= options.getUnitsPerSecond();
         }
-        // Truncate nanos to required units (millis / micros) and add to value
-        return value + (nanos / options.getNanosPerUnit());
+        // Truncate nanos to required units (millis / micros)
+        long truncatedNanos = nanos / options.getNanosPerUnit();
+        return getValueWithNanos(enableMicroPrecision, value, truncatedNanos);
+    }
+
+    private static long getSecondsInRequiredUnits(boolean enableMicroPrecision, long secondsWithBase, long unitsPerSecond)
+    {
+        try {
+            return enableMicroPrecision ? Math.multiplyExact(secondsWithBase, unitsPerSecond) : secondsWithBase * unitsPerSecond;
```

Comment:
Rewrite the code as
if (!enableMicroPercision) {
    // This can overflow/underflow, but to maintain backward compatibility this is not bounds checked.
    return regular multiplication
}
else {
    // Overflow/underflow is detected and the code will raise error.
   // other code
}